### ISSUE TITLE
[FEATURE] CPU / Memory hotplug support

### ISSUE THREAD

**Issue by innobead** [harvester/harvester#5835]
**Is your feature request related to a problem? Please describe.**

KubeVirt 1.1 supports CPU / Memory hotplug to dynamically update cpu / memory claims for VMs at runtime. Harvester should leverage this feature to have flexible and efficient resource management. 

**Additional context**

- https://kubevirt.io/user-guide/compute/cpu_hotplug/
- https://kubevirt.io/user-guide/operations/memory_hotplug/
- https://kubevirt.io/user-guide/operations/node_overcommit/

> Memory hotplug was introduced in KubeVirt version 1.1, enabling the dynamic resizing of the amount of memory available to a running VM.

**Comment by FrankYang0529** [comment]
Some prerequisite for this feature:

* Bump KubeVirt to v1.2, because there is some redesign for hotplug feature (https://github.com/kubevirt/kubevirt/issues/11900#issuecomment-2111775208)
* UI Enhancement:
  * UI set memory to `spec.template.spec.domain.resources.limits.memory`, but this feature can't have value in this field.
  * If we manually remove `spec.template.spec.domain.resources.limits.memory`, the dashboard can't show correct memory.
  <img width="974" alt="Screenshot 2024-07-17 at 4 21 37 PM" src="https://github.com/user-attachments/assets/3c54c2bb-87a5-4978-b730-110db0104df9">

**Comment by FrankYang0529** [comment]
To implement the feature, we may need to change following parts:

1. VM memory setting on UI.
When we set a VM memory on UI, it sets the value to the `resources.limits.memory` field. However, this blocks memory hotplug, because the limitation is set. Also, the `resources` field is not live upgradable, so we can't send a command with two patches for `resources.limits.memory` and `memory.guest`. We need to set the initial memory value to the `memory.guest` field.

2. Add two fields `Max CPU Hotplug` and `Max Memory Hotplug`.
The value in VM `resources.limits` is set to pod `resources.limits`. If we don't set `resources.limits` on VM, the pod may use whole node resources. We can have two new fields `Max CPU Hotplug` and `Max Memory Hotplug` on UI, so we can use these values as `resources.limits`.

3. Set `MacAddress` to annotation.
When we create a new VM, a [controller](https://github.com/harvester/harvester/blob/3b9a3e36125de9b4c500703803bd7807e89a7bf5/pkg/controller/master/virtualmachine/vmi_network_controller.go#L49-L86) gets `MacAddress` from VMI and sets it to the VM. The `MacAddress` field is non-live-upgradable, so it makes a new VM needs to restart to enable cpu / memory hotplug feature. We need to move the `MacAddress` field to annotation to avoid this.

4. Add an action to modify cpu / memory.
When we use "Edit Config" page to modify VM, it sends a request to steve server. This request is not handled by Harvester and it uses "update" to update the CR. However, the cpu / memory hotplug needs to use "patch" to update the CR. We need to add an action and do the patch for cpu / memory hotplug.

**Comment by w13915984028** [comment]
@FrankYang0529  Regarding https://github.com/harvester/harvester/issues/5835#issuecomment-2500010234

As the MAC has already been backfilled to VM's spec in general, does it mean all such VMs on existing cluster need to resart to get cpu/memory hotplug? thanks.

>Set MacAddress to annotation.
When we create a new VM, a [controller](https://github.com/harvester/harvester/blob/3b9a3e36125de9b4c500703803bd7807e89a7bf5/pkg/controller/master/virtualmachine/vmi_network_controller.go#L49-L86) gets MacAddress from VMI and sets it to the VM. The MacAddress field is non-live-upgradable, so it makes a new VM needs to restart to enable cpu / memory hotplug feature. We need to move the MacAddress field to annotation to avoid this.

**Comment by w13915984028** [comment]
Another question: how to deal with `ResourceQuota` when hot plug request is sent?

**Comment by w13915984028** [comment]
@FrankYang0529 
Another question: if the VM has expanded the guest to max allowed cpu and memory, namely no `hot-plug` room, should `hot-plug` menu still be visiable?  thanks.

https://github.com/harvester/harvester/blob/96302ffb12623c08a309a399162df68ca2c852aa/pkg/api/vm/formatter.go#L410

**Comment by FrankYang0529** [comment]
For single node case, it cannot migrate VM successfully now. Create another issue to track it. https://github.com/harvester/harvester/issues/8774

**Comment by albinsun** [comment]
Hi @m-ildefons ,
Please help to verify this issue, thank you.

**Comment by TachunLin** [comment]
### HEP
* https://github.com/harvester/harvester/blob/master/enhancements/20250107-cpu-memory-hotplug.md#test-plan


### UI page covered
1. Create vm -> check CPU and memory hotplug
2. Edit CPU and Memory of VM -> Edit panel
3. VM -> Show config -> Check maximum CPU and Memory of VM
4. VM -> Edit config -> Edit maximum CPU and Memory of VM
5. Settings -> max-hotplug-ratio


### Test plans 

### Test plans 

1. Create new vm with CPU and memory Hotplug checked
2. Create new vm without CPU and memory Hotplug checked
3. Set the customized maximum CPU and memory value to new VM 
4. Check the maximum CPU and memory value on `Show Configuration` panel
5. Edit VM and adjust the current CPU and Memory
6. The required config attribute can be added to the VM yaml content
7. (Corner) Test the edge case value input of the maximum CPU and memory value
8. Edit the `max-hotplug-ratio` value in setting and test with the maximum CPU and memory value in VM
9. Test the default value of `max-hotplug-ratio
10. (Corner) Test the edge case value input of the `max-hotplug-ratio` value
11. Dynamic modify the `sockets` of running VM, test the rule to bigger than current `sockets` and smaller than the `maxSockets`
12. Dynamic modify the `memory` of running VM, test the rule to bigger than current `memory` and smaller than the `maxGuest`
13. Dynamic modify the `sockets` of running VM, test the rule to smaller than current `sockets`
14. Dynamic modify the `memory` of running VM, test the rule to smaller than current `memory`
15. Dynamic modify the `maxSockets` and `maxGuest` of running VM
16. Disable the CPU and memory Hotplug of the running VM
17. Create new vm without CPU and memory Hotplug checked and enable it afterward
18. Test major vm operations when set the maximum CPU and memory value
19. Test major vm operations after dynamic edit the `socket` and `memory` value

**Comment by TachunLin** [comment]
Verified **fixed** on `v1.6.0-rc5`. Close this issue.

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Create new vm with CPU and memory Hotplug checked $~~$</summary>

1. Can enable the `Enable CPU and memory hotplug`, and automatically provide the maximum value according to the default ratio `4` 
    <img width="1600" height="783" alt="Image" src="https://github.com/user-attachments/assets/4dde40c5-91c7-4beb-90c7-828e1f32f2b5" />

2. Can create VM into running state
  
    <img width="1599" height="784" alt="Image" src="https://github.com/user-attachments/assets/f89b47d0-749d-4945-9628-fff87644895b" />

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Create new vm without CPU and memory Hotplug checked $~~$</summary>


1. Can create VM into running state
  

  <img width="1600" height="784" alt="Image" src="https://github.com/user-attachments/assets/c09c8a5c-6bda-40fa-85e0-d4b2ac84e05b" />


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Set the customized maximum CPU and memory value to new VM  $~~$</summary>


1. Can set customized maximum CPU and memory value to new VM 

<img width="1600" height="795" alt="Image" src="https://github.com/user-attachments/assets/96c8cefb-65f4-44f5-a04a-8b7678ddd87d" />

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Check the maximum CPU and memory value on Show Configuration panel $~~$</summary>

1. Can display the correct maximum CPU and memory value and option checked state
    
    <img width="1600" height="783" alt="Image" src="https://github.com/user-attachments/assets/1554ffae-8c15-4408-a6ac-541d4168abf4" />

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Edit VM and adjust the current CPU and Memory $~~$</summary>

1. Can adjust the current cpu and memory and vm can restart into running state

  https://github.com/user-attachments/assets/5883551b-1620-4bba-ba6c-74ca0fa4a5fd
  
 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ The required config attribute can be added to the VM yaml content $~~$</summary>

1. Confirm the `domain.cpu.sockets` and `domain.cpu.maxSockets`

  <img width="1135" height="699" alt="Image" src="https://github.com/user-attachments/assets/9bab63a9-8eb7-4912-9ff6-0468c400f44f" />

2. Confirm the `domain.memory.guest` and `domain.memory.guest`

  <img width="1598" height="783" alt="Image" src="https://github.com/user-attachments/assets/003ce9e9-bd93-4398-8578-3c197a0f8730" />

 
3. Confirm the `resources.limits.cpu` and `resources.limits.memory`

  <img width="1600" height="787" alt="Image" src="https://github.com/user-attachments/assets/8ae6cc10-d5fc-455e-8858-8ec36d11e79e" />

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ (Corner) Test the edge case value input of the maximum CPU and memory value $~~$</summary>

1. Can block negative integer value of maximum cpu and memory

  <img width="1600" height="785" alt="Image" src="https://github.com/user-attachments/assets/23d37922-a2a0-436b-82e2-5703b434487d" />

1. Can block zero integer value of maximum cpu and memory

  <img width="1600" height="781" alt="Image" src="https://github.com/user-attachments/assets/2087dd99-678f-4d27-ac76-cfd9da50e07d" />
  
  <img width="1600" height="784" alt="Image" src="https://github.com/user-attachments/assets/74ccb12b-8547-46f4-bf0d-92016f5b4261" />
  
  https://github.com/user-attachments/assets/596b11ea-77fc-44d5-9578-d92bb17d02b2

1. Set extra large value of maximum cpu and memory, no limit here but it will be blocked by the Golang language integer range
  

  <img width="1600" height="786" alt="Image" src="https://github.com/user-attachments/assets/8ccb5c2a-5c4b-46b3-912a-f54aacf29105" />

  https://github.com/user-attachments/assets/eae348db-6f45-4c5c-bcee-4d71c33a9fa2

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Edit the max-hotplug-ratio value in setting and test with the maximum CPU and memory value in VM $~~$</summary>

1. We can change the hotplug-ratio to max `20` and it can be taken effect

  <img width="960" height="164" alt="Image" src="https://github.com/user-attachments/assets/7ee344b3-028c-4c42-8d4d-0962b2f350c6" />
  
  <img width="1596" height="788" alt="Image" src="https://github.com/user-attachments/assets/cee8f10c-01df-4fdc-81a8-ff68ddac1c1b" />
  
  <img width="1600" height="784" alt="Image" src="https://github.com/user-attachments/assets/3eb46916-be67-4343-83a4-6e298b1b43ad" />

1. We can change the hotplug-ratio to min `1` and it can be taken effect

  <img width="851" height="149" alt="Image" src="https://github.com/user-attachments/assets/66fb2491-c4ed-486e-9821-6c5cde0c8568" />
  
  <img width="1341" height="491" alt="Image" src="https://github.com/user-attachments/assets/c57ead2a-8c2f-40b8-a879-530cb7e39f15" />
  
  <img width="1597" height="781" alt="Image" src="https://github.com/user-attachments/assets/d0acdda6-befb-46dc-85c0-6293eb420ddf" />

  
 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test the default value of `max-hotplug-ratio $~~$</summary>

1. Click the `Use the default value`, the ratio will change back to the default value `4`
  

  <img width="1596" height="789" alt="Image" src="https://github.com/user-attachments/assets/014936d0-7d25-4d57-9683-3a29a6d065c5" />

1. And the ratio can be taken effect on vm

  <img width="1597" height="786" alt="Image" src="https://github.com/user-attachments/assets/19774cc5-705f-4b3d-a436-13fcbefce696" />

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ (Corner) Test the edge case value input of the max-hotplug-ratio value $~~$</summary>

1. Set ratio value to negative number would be blocked
  

  <img width="1600" height="798" alt="Image" src="https://github.com/user-attachments/assets/94a209e2-edfc-47e8-9198-e56436f59e84" />


1. Set the ratio value to zero would be blocked

  <img width="1598" height="782" alt="Image" src="https://github.com/user-attachments/assets/83e219c6-01df-48c1-915f-367bc7f21d25" />

1. Set ratio value larger than 20 would be blocked

  <img width="1600" height="786" alt="Image" src="https://github.com/user-attachments/assets/52341167-1cd8-4126-a61f-81db1bfa7903" />

1. Set the ratio value to decimal number would be blocked

  <img width="1600" height="790" alt="Image" src="https://github.com/user-attachments/assets/a346e978-d585-4961-9bf3-aeccb96dfce9" />

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Dynamic modify the sockets of running VM, test the rule to bigger than current sockets and smaller than the maxSockets $~~$</summary>


1. Can edit the CPU `socket` to larger than the current socket and smaller than the `maxSockets`
2. VM won't restart
3. VM can successfully migrate to the specific node
4. After migrated, the VM can display the updated CPU socket number
5. Access to the vm, the CPU socket actually changed 

<img width="1600" height="782" alt="Image" src="https://github.com/user-attachments/assets/16eb4963-3590-4e4a-8473-ab4ec3532404" />

<img width="1600" height="786" alt="Image" src="https://github.com/user-attachments/assets/b62eb6b6-519b-4715-a0ed-bcc2e6746319" />


  ```
  edit-cpu-memory-dynamic:~ # lscpu
  Architecture:             x86_64
    CPU op-mode(s):         32-bit, 64-bit
    Address sizes:          42 bits physical, 48 bits virtual
    Byte Order:             Little Endian
  CPU(s):                   5
  ```

  https://github.com/user-attachments/assets/3c8fd939-4411-4300-88ec-9e67ab4f7757

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Dynamic modify the memory of running VM, test the rule to bigger than current memory and smaller than the maxGuest $~~$</summary>


1. Can edit the `memory` to larger than the current socket and smaller than the `maxSockets`
2. VM won't restart
3. VM can successfully migrate to the specific node
4. After migrated, the VM can display the updated memory number

  <img width="1600" height="782" alt="Image" src="https://github.com/user-attachments/assets/92bf4a8b-fd97-4982-b5c6-5c85db037dcd" />
  
  https://github.com/user-attachments/assets/918174e5-16fa-47f6-b332-e7a94ab0ff71


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Dynamic modify the sockets of running VM, test the rule to smaller than current sockets $~~$</summary>


1. Can edit the `sockers` to smaller than the current socket
2. Click save
3. It would prompt the message `Restart action is required for the virtual machine configuration change to take effect`

  <img width="1600" height="786" alt="Image" src="https://github.com/user-attachments/assets/d107992f-2464-4ea8-91d5-54f1c79565df" />

  https://github.com/user-attachments/assets/2b898707-b3e7-4dab-871a-98f46f9ef847
  
4. After VM restarted, we can find the cpu have changed to the updated value

  ```
  opensuse@edit-cpu-memory-dynamic:~> lscpu
  Architecture:             x86_64
    CPU op-mode(s):         32-bit, 64-bit
    Address sizes:          42 bits physical, 48 bits virtual
    Byte Order:             Little Endian
  CPU(s):                   4
    On-line CPU(s) list:    0-3
  
  ```

 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Dynamic modify the memory of running VM, test the rule to smaller than current memory $~~$</summary>


1. Can edit the `memory` to smaller than the current memory
2. Click save
3. It would prompt the message `Restart action is required for the virtual machine configuration change to take effect`


  <img width="1598" height="784" alt="Image" src="https://github.com/user-attachments/assets/94013f18-9692-4ee2-b1c4-1f5c21a04c17" />
  
  
  https://github.com/user-attachments/assets/66985501-63a5-421a-ad90-0547b24f1223

  
4. After VM restarted, we can find the memory have changed to the updated value

  ```
  opensuse@edit-cpu-memory-dynamic:~> free --giga
                 total        used        free      shared  buff/cache   available
  Mem:               6           0           5           0           0           5
  Swap:              0           0           0
  ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Dynamic modify the maxSockets and maxGuest of running VM $~~$</summary>


1. Can edit the `maxSockets` and `maxGuest` of the running VM
2. Click save it would prompt to restart the VM
3. After VM restarted, we can find the `maxSockets` and `maxGuest` value have been changed

  <img width="1600" height="779" alt="Image" src="https://github.com/user-attachments/assets/246a5c7b-87ee-4b94-93ff-29e4deff5169" />
  
  <img width="1599" height="783" alt="Image" src="https://github.com/user-attachments/assets/09b81a43-de44-4bb8-8263-f61c61b4915c" />
  
  https://github.com/user-attachments/assets/4ba84eb7-eeb3-49f7-b921-53441d851e61

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Disable the CPU and memory Hotplug of the running VM $~~$</summary>

1. After disable the CPU and memory hogtplug, it require to restart VM
2. Can restart vm into running state, and confirm the option have disabled

    <img width="1600" height="787" alt="Image" src="https://github.com/user-attachments/assets/df657ac1-438f-4cab-9e7b-9154c8a0fae6" />
  
  https://github.com/user-attachments/assets/b866e7dc-2fab-4229-a42c-bd31be267546

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Create new vm without CPU and memory Hotplug checked and enable it afterward $~~$</summary>

1. Create a regular vm without CPU and memory hotplug enabled
2. After the VM is running, edit the VM and check the hotplug option
3. It require to restart VM
4. After VM restarted into running
5. The CPU and memory hotplug is checked and maximum values have been set

  <img width="1600" height="787" alt="Image" src="https://github.com/user-attachments/assets/cd0d0ee3-f971-4aed-85b5-346858a01050" />
  
  
  https://github.com/user-attachments/assets/1a93c2df-89b4-4105-a013-9a224508251f


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test major vm operations when set the maximum CPU and memory value
 $~~$</summary>

1.  When a VM have set the maximum CPU and memory value, test the following VM operations

  - ✅ Restart
  - ✅ Stop
  - ✅ Start
  - ✅ Pause/Unpause
  - ✅ Take backup
  - ✅ Take snapshot
  - ✅ Clone
  - ✅ Migrate
  - ✅ Add volume
  - ✅ Delete 

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test major vm operations after dynamic edit the socket and memory value $~~$</summary>

1.  When a VM have dynamically edit the maximum CPU and memory value, test the following VM operations

  - ✅ Restart
  - ✅ Stop
  - ✅ Start
  - ✅ Pause/Unpause
  - ✅Take backup
  - ✅ Take snapshot
  - ✅ Clone
  - ✅ Migrate
  - ✅ Add volume
  - ✅ Delete 


 </details>


### Test Information
* Test Environment: Three nodes bare-metal machines
* Harvester version: `v1.6.0-rc5`
* Rancher version: v2.12.0
* Harvester UI extension: `v1.6.0-rc5`

### Verify Steps
<details><summary> CPU and Memory hotplug feature </summary>

 Refer to: https://github.com/harvester/harvester/issues/5835#issuecomment-3175966364

</details>
